pkg/api/controller_test.go

```diff
@@ -391,6 +391,94 @@ func TestBasicAuth(t *testing.T) {
 	})
 }
 
+func TestClientInterrupts(t *testing.T) {
+	Convey("Interrupted blob upload", t, func() {
+		port := getFreePort()
+		baseURL := getBaseURL(port, false)
+		config := api.NewConfig()
+		config.HTTP.Port = port
+
+		c := api.NewController(config)
+		dir, err := ioutil.TempDir("", "oci-repo-test")
+		if err != nil {
+			panic(err)
+		}
+
+		defer os.RemoveAll(dir)
+		c.Config.Storage.RootDirectory = dir
+		go func() {
+			// this blocks
+			if err := c.Run(); err != nil {
+				return
+			}
+		}()
+
+		// wait till ready
+		for {
+			_, err := resty.R().Get(baseURL)
+			if err == nil {
+				break
+			}
+			time.Sleep(100 * time.Millisecond)
+		}
+
+		defer func() {
+			ctx := context.Background()
+			_ = c.Server.Shutdown(ctx)
+		}()
+
+		blob := make([]byte, 75*1024*1024)
+
+		digest := godigest.FromBytes(blob).String()
+
+		resp, err := resty.R().Post(baseURL + "/v2/" + AuthorizedNamespace + "/blobs/uploads/")
+		So(err, ShouldBeNil)
+		So(resp, ShouldNotBeNil)
+		So(resp.StatusCode(), ShouldEqual, 202)
+
+		loc := resp.Header().Get("Location")
+		splittedLoc := strings.Split(loc, "/")
+		sessionID := splittedLoc[len(splittedLoc)-1]
+
+		ctx := context.Background()
+		timeout := 1 * time.Second
+		ctx, cancel := context.WithTimeout(ctx, timeout)
+
+		_, _ = resty.R().
+			SetHeader("Content-Length", fmt.Sprintf("%d", len(blob))).
+			SetHeader("Content-Type", "application/octet-stream").
+			SetQueryParam("digest", digest).
+			SetBody(blob).
+			SetContext(ctx).
```

Comment:
I can not simulate a client interrupt... it behaves different on different machines.
For example this test passes all the time on my machine.

If you have any suggestion, let me know.

Reply:
Push a few blobs and go to sleep before it finishes and the other goroutine interrupts it.

Reply:
Ok, I finally found a way to tell when the upload has started, and then interrupt. Thanks!